frontend/javascripts/admin/user/permissions_and_teams_modal_view.tsx

```diff
@@ -107,63 +85,55 @@ class PermissionsAndTeamsModalView extends React.PureComponent<TeamRoleModalProp
       previousPermission = PERMISSIONS.datasetManager;
     }
 
-    return previousPermission !== this.state.selectedPermission;
-  };
-  handleUpdatePermissionsAndTeams = () => {
-    if (this.didPermissionsChange()) {
-      const user = getSingleUserMaybe(this.props);
+    return previousPermission !== selectedPermission;
+  }
+
+  function handleUpdatePermissionsAndTeams() {
+    if (didPermissionsChange()) {
+      const user = getSingleUserMaybe();
 
       if (user) {
         const userName = `${user.firstName} ${user.lastName}`;
         let message = messages["users.revoke_all_permissions"];
 
-        if (this.state.selectedPermission === PERMISSIONS.admin) {
+        if (selectedPermission === PERMISSIONS.admin) {
           message = messages["users.set_admin"];
         }
 
-        if (this.state.selectedPermission === PERMISSIONS.datasetManager) {
+        if (selectedPermission === PERMISSIONS.datasetManager) {
           message = messages["users.set_dataset_manager"];
         }
 
-        Modal.confirm({
+        modal.confirm({
           title: messages["users.change_permissions_title"],
           content: message({
             userName,
           }),
-          onOk: this.setPermissionsAndTeams,
+          onOk: setPermissionsAndTeams,
         });
       }
     } else {
-      this.setPermissionsAndTeams();
+      setPermissionsAndTeams();
     }
-  };
-  setPermissionsAndTeams = () => {
-    const newUserPromises = this.props.users.map((user) => {
-      if (this.props.selectedUserIds.includes(user.id)) {
-        const newTeams = Utils.values(this.state.selectedTeams);
-        const newUser = Object.assign({}, user, {
-          teams: newTeams,
-        });
+  }
+
+  function setPermissionsAndTeams() {
+    const newUserPromises = users.map((user) => {
+      if (selectedUserIds.includes(user.id)) {
+        const newTeams = Utils.values(selectedTeams);
+        let permissions = { isAdmin: false, isDatasetManager: false };
```

Comment:
Instead of manipulating the `newUser` (read-only attributes) variable directly, we figure out the permissions first and then return a new user object. Typescript likes this a lot more.